File: 128. Longest Consecutive Sequence.py
'''
Given an unsorted array of integers, find the length of the longest consecutive elements sequence.

Your algorithm should run in O(n) complexity.

Example:

Input: [100, 4, 200, 1, 3, 2]
Output: 4
Explanation: The longest consecutive elements sequence is [1, 2, 3, 4]. Therefore its length is 4.
'''

#搭配leetcode 128一起服用
#這題是求sequence, 所以不必連續字串
#思路: 先用set來消除重複元素, 找longest consecutive sequence 只需一個元素代表即可, 此題概念首尾貪吃蛇
#若一個元素在consecutive sequence上, 往前往後查找其他成員值, 一定可以把全部元素找出來, 
#這此題, 把找出來的元素從set拿掉=> 因為每個元素只會被pop or remove 1 次, 所以time complexity O(n)
class Solution:
    def longestConsecutive(self, num):
        num=set(num)
        maxLen=0
        while num:
            n=num.pop()
            i=n+1 #找是否有元素 =  n+ 1
            l1=0
            l2=0
            while i in num:
                num.remove(i) #set.remove() O(1) time complexity
                i+=1
                l1+=1
            i=n-1  #換找是否有 n-1 元素在裡面
            while i in num:
                num.remove(i)
                i-=1
                l2+=1
            maxLen=max(maxLen,l1+l2+1)  #+1, 記得把自己加進去
        return maxLen


#自己重寫, 刷題用這個
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        new_nums = set(nums)
        max_len = 0
        while new_nums:
            l1 = 0
            l2 = 0
            n = new_nums.pop()
            i = n + 1
            while i in new_nums:
                new_nums.remove(i)
                l1 += 1
                i += 1
            i = n - 1
            while i in new_nums:
                new_nums.remove(i)
                l2 += 1
                i -= 1
            max_len = max(max_len, l1+l2+1)
        return max_len

File: leetcode_cpp/128. Longest Consecutive Sequence.cpp
// time complexity O(n), 參照 pyrhon 解法
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> set;
        for (auto num: nums){
            set.insert(num);
        }
        int max_len {0};
        while (!set.empty()){
            int temp = *set.begin(); // 不能 temp = *--set.end() => 因為set.end() 是 forward iterators
            set.erase(temp);
            int l {0};
            int r {0};
            int i = temp + 1;
            while (set.find(i) != set.end()){
                set.erase(i);
                i += 1;
                r += 1;   
            }
            i = temp - 1;
            while (set.find(i) != set.end()){
                set.erase(i);
                i -= 1;
                l += 1;
            }
        max_len = max(max_len, l + 1 + r);
        }
        return max_len;
    }
};

// set.end() return type is a forward iterator, You cannot use decrement operator with forward iterators
